File: 2025-09/2025-09-25/2nd_largest.py
def second_largest(arr):

    return sorted(set(arr))[-2]

print(second_largest([1, 2, 3, 4]))
print(second_largest([20, 139, 94, 67, 31]))
print(second_largest([2, 3, 4, 6, 6]))
print(second_largest([10, -17, 55.5, 44, 91, 0]))
print(second_largest([1, 0, -1, 0, 1, 0, -1, 1, 0]))

# --- TEST SUITE ---

tests_text = '''
1. second_largest([1, 2, 3, 4]) should return 3.
2. second_largest([20, 139, 94, 67, 31]) should return 94.
3. second_largest([2, 3, 4, 6, 6]) should return 4.
4. second_largest([10, -17, 55.5, 44, 91, 0]) should return 55.5.
5. second_largest([1, 0, -1, 0, 1, 0, -1, 1, 0]) should return 0.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 09 September/09-25/2ndLargest.js
function secondLargest(arr) {

  return Array.from(new Set(arr))
    .sort((a, b) => a - b)
    .at(-2);
}

console.log(secondLargest([1, 2, 3, 4]));
console.log(secondLargest([20, 139, 94, 67, 31]));
console.log(secondLargest([2, 3, 4, 6, 6]));
console.log(secondLargest([10, -17, 55.5, 44, 91, 0]));
console.log(secondLargest([1, 0, -1, 0, 1, 0, -1, 1, 0]));

// --- TEST SUITE ---

const testsText = `
1. secondLargest([1, 2, 3, 4]) should return 3.
2. secondLargest([20, 139, 94, 67, 31]) should return 94.
3. secondLargest([2, 3, 4, 6, 6]) should return 4.
4. secondLargest([10, -17, 55.5, 44, 91, 0]) should return 55.5.
5. secondLargest([1, 0, -1, 0, 1, 0, -1, 1, 0]) should return 0.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);